File: src/fableoutput/components/LoadData.js
import { class_type } from "../fable_modules/fable-library-js.4.19.0/Reflection.js";
import { createElement } from "react";
import React from "react";
import { useReact_useContext_37FA55CF, useFeliz_React__React_useState_Static_1505 } from "../fable_modules/Feliz.2.6.0/React.fs.js";
import { Pages, UserData } from "../ReactContext.js";
import { createObj } from "../fable_modules/fable-library-js.4.19.0/Util.js";
import { singleton, delay, toList } from "../fable_modules/fable-library-js.4.19.0/Seq.js";
import { Interop_reactApi } from "../fable_modules/Feliz.2.6.0/Interop.fs.js";
import { PromiseBuilder__Delay_62FBFDE1, PromiseBuilder__Run_212F1D4B } from "../fable_modules/Fable.Promise.3.2.0/Promise.fs.js";
import { promise } from "../fable_modules/Fable.Promise.3.2.0/PromiseImpl.fs.js";
import { UserData as UserData_1 } from "../MockData.js";
import { Pages as Pages_1 } from "../Types.js";
import { some } from "../fable_modules/fable-library-js.4.19.0/Option.js";
import { ofArray } from "../fable_modules/fable-library-js.4.19.0/List.js";

export class LoadData {
    constructor() {
    }
}

export function LoadData_$reflection() {
    return class_type("App.LoadData", undefined, LoadData);
}

export function LoadData_Main() {
    let elems;
    const patternInput = useFeliz_React__React_useState_Static_1505(false);
    const setLoading = patternInput[1];
    const loading = patternInput[0];
    const dataCtx = useReact_useContext_37FA55CF(UserData);
    const pageCtx = useReact_useContext_37FA55CF(Pages);
    return createElement("button", createObj(ofArray([["className", "btn btn-primary"], (elems = toList(delay(() => (loading ? singleton(createElement("div", {
        className: "loading loading-spinner",
        style: {
            width: 20,
            height: 20,
        },
    })) : singleton("Load mock data")))), ["children", Interop_reactApi.Children.toArray(Array.from(elems))]), ["onClick", (_arg) => {
        let pr_1;
        const pr = PromiseBuilder__Run_212F1D4B(promise, PromiseBuilder__Delay_62FBFDE1(promise, () => {
            setLoading(true);
            return (new Promise(resolve => setTimeout(resolve, 1000))).then(() => {
                dataCtx.setData(UserData_1);
                setLoading(false);
                pageCtx.setData(new Pages_1(1, []));
                return Promise.resolve();
            });
        }));
        pr_1 = (pr.catch((ex) => {
            console.error(some(ex));
            setLoading(false);
        }));
        void pr_1;
    }]])));
}

File: src/components/LoadData.fs
namespace App

open Feliz
open Fable.Core.JsInterop

type LoadData =
    [<ReactComponent>]
    static member Main() =
        let loading, setLoading = React.useState(false)
        let dataCtx = React.useContext(App.ReactContext.UserData)
        let pageCtx = React.useContext(App.ReactContext.Pages)
        let uploadRef = React.useInputRef()

        let setData = fun (fileName: string) (tsvValues: string[][]) ->
            promise {
                setLoading true
                let userData: UserData = {
                    FileName = fileName
                    Data = tsvValues
                }
                userData |> Some |> dataCtx.setData
                setLoading false
                pageCtx.setData Pages.SelectIdCol
            }
            |> Promise.catch (fun ex ->
                Browser.Dom.console.error ex
                setLoading false
            )


        React.fragment [
            Html.div [
                prop.className "flex flex-col md:flex-row gap-2 justify-center items-center"
                prop.children [
                    Html.input [
                        prop.ref uploadRef
                        prop.className "file-input file-input-primary"
                        prop.type'.file
                        prop.onChange(fun (file: Browser.Types.File) ->
                            if isNull file then
                                ()
                            else
                                promise {
                                    let! (text: string) = file?text()
                                    let values =
                                        text.Split([| "\r\n"; "\n" |], System.StringSplitOptions.None)
                                        |> Array.map (fun line ->
                                            line.Split([| "\t" |], System.StringSplitOptions.None)
                                            |> Array.map (fun value -> value.Trim())
                                        )
                                    do! setData file.name values
                                }
                                |> Promise.catch (fun ex ->
                                    Browser.Dom.console.error ex
                                    setLoading false
                                )
                                |> Promise.start
                        )
                    ]
                    Html.button [
                        prop.className "btn btn-outline";
                        prop.children [
                            Html.text "Load mock data"
                        ]
                        prop.onClick (fun _ ->
                            setData MockData.UserData.FileName MockData.UserData.Data
                            |> Promise.start
                        )
                    ]

                ]
            ]
            if loading then
                Basic.LoadingModal("Loading data...")
        ]